### PR TITLE
 #167727640 User(manager) can Reject a Request

### PR DESCRIPTION
**Pull request by ChiamakaObitube**
#### What does this PR do?
Ensures that a user(manager) can reject a trip request.

`PATCH /api/v1/request/reject`

#### Description of Task to be completed?
Given that a user is a manager when he logs in,
he should be able to reject a trip request.
Only a manager has the capacity to reject a trip request

#### How should this be manually tested?
* Clone the repo and `cd` into the project directory
* Run `yarn install` to install the project's dependency
* Run `yarn start` to start the app
Login with a manager credential (email and password)
Send a PATCH request to  `localhost:3000/api/v1/request/reject`

#### What are the relevant pivotal tracker stories?
[#167727640](https://www.pivotaltracker.com/story/show/167727640)

#### Any background context you want to add?
None

### CODE REVIEW

File: src/models/User.js [hunk 1]
```diff
@@ -185,8 +185,8 @@ export default (sequelize, DataTypes) => {
   {
     classMethods: {
       associate: models => {
-        User.hasMany(models.Requests, {
-          foreignKey: 'userId',
+        User.hasMany(models.Request, {
+          foreignKey: 'UserId',
```
Comment (chrismeeky) [inline src/models/User.js]: why did you change the foreign key?

File: src/routes/requestRoute.js [hunk 2]
```diff
@@ -12,6 +12,11 @@ const requestRoutes = app => {
     Authorization.checkToken,
     Validate.validateUserInput,
     RequestController.bookAReturnTrip);
+
+  app.patch('/api/v1/request/reject/:id',
```
Comment (chrismeeky) [inline src/routes/requestRoute.js]: We don't pass ID as a URL parameter. Pass it in the request body

File: src/seeders/03-demo-request.js [hunk 3]
```diff
@@ -7,8 +7,9 @@ module.exports = {
       flightDate: '2019-06-21',
       returnDate: '2019-03-21',
       accommodationId: '2125be7b-f1f1-4f0a-af86-49c657870b5c',
-      userId: '3821b930-ce48-4ac8-9ddf-ee3bf7980d08',
+      UserId: '3821b930-ce48-4ac8-9ddf-ee3bf7980d08',
```
Comment (chrismeeky) [inline src/seeders/03-demo-request.js]: why did you change userId to UserId ?

File: src/seeders/03-demo-request.js [hunk 4]
```diff
@@ -19,8 +20,9 @@ module.exports = {
       flightDate: '2019-06-21',
       returnDate: '2019-03-21',
       accommodationId: '9c41e609-7a30-4211-9d10-146a9c54ee74',
-      userId: '96dc6b6d-7a77-4322-8756-e22f181d952c',
+      UserId: '96dc6b6d-7a77-4322-8756-e22f181d952c',
```
Comment (chrismeeky) [inline src/seeders/03-demo-request.js]: ditto

File: src/seeders/03-demo-request.js [hunk 5]
```diff
@@ -31,8 +33,9 @@ module.exports = {
       flightDate: '2019-06-21',
       returnDate: '2019-03-21',
       accommodationId: '2125be7b-f1f1-4f0a-af86-49c657870b5c',
-      userId: '79ddfd3b-5c83-4beb-815e-55b1c95230e1',
+      UserId: '79ddfd3b-5c83-4beb-815e-55b1c95230e1',
```
Comment (chrismeeky) [inline src/seeders/03-demo-request.js]: ditto

File: src/controllers/RequestController.js [hunk 6]
```diff
@@ -21,7 +21,7 @@ class RequestController {
     try {
       const { id } = req.decoded;
       const { body } = req;
-      const { dataValues } = await Request.create({ ...body, userId: id });
+      const { dataValues } = await Request.create({ ...body, UserId: id });
```
Comment (chrismeeky) [inline src/controllers/RequestController.js]: why did you change userId to UserId

File: src/migrations/03-create-requests.js [hunk 7]
```diff
@@ -5,7 +5,7 @@ module.exports = {
       type: Sequelize.UUID,
       defaultValue: Sequelize.UUIDV4
     },
-    userId: {
+    UserId: {
```
Comment (chrismeeky) [inline src/migrations/03-create-requests.js]: why did you change userId to UserId

File: src/routes/requestRoute.js [hunk 8]
```diff
@@ -12,6 +12,11 @@ const requestRoutes = app => {
     Authorization.checkToken,
     Validate.validateUserInput,
```
Comment (chrismeeky) [inline src/routes/requestRoute.js]: is this necessary? since you are not validating any input ?

File: src/controllers/RequestController.js [hunk 9]
```diff
@@ -66,6 +66,43 @@ class RequestController {
...
+  * @param {object} res - HTTP Response object
+  * @return {res} res - HTTP Response object
+  * @memberof RequestController
+  */
+  static async rejectRequest(req, res) {
+    try {
+      const { role } = req.decoded;
+      const { id } = req.params;
+
+      if (role === 'Manager') {
```
Comment (chrismeeky) [inline src/controllers/RequestController.js]: Your logic doesn't check if request is from Manager's direct report.

Comment (ChiamakaObitube) [reply src/controllers/RequestController.js]: Hello Chris, thank you for the feedback. The logic is a condition that checks if the user is a manager, he is able to reject a request. Outside the if statement on line 97 returns a statement if the user is not a manager.

File: src/models/User.js (hunk 1 above)
Comment (ChiamakaObitube) [reply src/models/User.js]: I changed it because Sequelize throws an error with camel-cased variables.

File: src/routes/requestRoute.js (hunk 2 above)
Comment (ChiamakaObitube) [reply src/routes/requestRoute.js]: The ID has to be passed into the request parameter so it can be easily fetched when searching for a particular request.

File: src/models/User.js (hunk 1 above)
Comment (chrismeeky) [reply src/models/User.js]: to solve the problem, there is an extra association in the User model. delete it

File: src/routes/requestRoute.js (hunk 2 above)
Comment (chrismeeky) [reply src/routes/requestRoute.js]: You can check the code base for previous routes. All id's were passed in the request body for security reasons. you can confirm from the TTL who effected the changes

File: src/controllers/RequestController.js (hunk 9 above)
Comment (chrismeeky) [reply src/controllers/RequestController.js]: Yes.I agree with the logic. But an extra logic has to be there to ensure that not every manager can reject request made by another manager's direct report. That's why there is a column called lineManager. You might want to refer to the PT story

File: src/routes/requestRoute.js (hunk 2 above)
Comment (ChiamakaObitube) [reply src/routes/requestRoute.js]: Okay. Thanks for the feedabck

File: src/controllers/RequestController.js (hunk 9 above)
Comment (chrismeeky) [reply src/controllers/RequestController.js]: I am working on a similar feature. You can check => https://www.pivotaltracker.com/story/show/167727641

Comment (ChiamakaObitube) [reply src/controllers/RequestController.js]: Yeah, but in this case, there is no defined organization structure that we can use here. How do I check if a particular user is another's user's line manager. From my understanding of the user roles, there are 5 user roles as provided in the PT board.

Comment (chrismeeky) [reply src/controllers/RequestController.js]: > How do I check if a particular user is another's user's line manager

Get all requests include User=> where user.line manager === managerId

File: src/controllers/RequestController.js (hunk 6 above)
Comment (Daymorelah) [reply src/controllers/RequestController.js]: Am waiting for your response to @chrismeeky 's question.

File: src/controllers/RequestController.js [hunk 10]
```diff
@@ -66,6 +66,39 @@ class RequestController {
...
+        return HelperMethods.clientError(res, 'You are not authorized to update this request',
+          401);
+      }
+      const requestExist = await Request.findByPk(req.body.id);
+      if (requestExist.dataValues.status === 'OPEN' && requestExist.dataValues.id) {
+        await requestExist.update({ status: 'REJECTED' });
+        return HelperMethods
+          .requestSuccessful(res, {
+            success: true,
+            message: 'Your request is rejected',
```
Comment (Daymorelah) [inline src/controllers/RequestController.js]: Good job @ChiamakaObitube. Is the manager rejecting his/her request? or is he/she rejecting someone else's request?. Please make your response message reflect what the request performed successfully.

File: src/models/Request.js [hunk 11]
```diff
@@ -46,12 +46,17 @@ export default (sequelize, DataTypes) => {
...
+      type: DataTypes.ENUM,
+      values: ['OPEN', 'REJECTED', 'APPROVED'],
+      defaultValue: 'OPEN'
     }
   });
 
   Request.associate = models => {
     Request.belongsTo(models.User, {
-      foreignKey: 'userId',
+      foreignKey: 'UserId',
```
Comment (Daymorelah) [inline src/models/Request.js]: why did you change all instances of `userId` to `UserId`

**Review by Daymorelah — CHANGES_REQUESTED** [review]

File: src/controllers/RequestController.js (hunk 10 above)
Comment (ChiamakaObitube) [reply src/controllers/RequestController.js]: Thank you @Daymorelah for the feedback. The manager is rejecting someone else's request.

File: src/models/Request.js (hunk 11 above)
Comment (ChiamakaObitube) [reply src/models/Request.js]: I changed it because Sequelize was throwing an error that 'UserId' does not exist when I try to run the app. So when I changed 'userId' to UserId', the error stopped.

File: src/models/User.js (hunk 1 above)
Comment (ChiamakaObitube) [reply src/models/User.js]: Okay.

File: src/controllers/RequestController.js (hunk 6 above)
Comment (ChiamakaObitube) [reply src/controllers/RequestController.js]: I changed it because Sequelize was throwing an error that 'UserId' does not exist when I try to run the app. So when I changed 'userId' to UserId', the error stopped.
